File: data_generation/src/data_generator.cpp
//
// Created by will on 04/04/24.
//

#include "../include/data_generator.h"

DataGenerator::DataGenerator(std::string filename): robot("manipulator") {
    this->filename = filename;
    this->ik_data.load_data(this->filename);
    this->ik_data_result = MasterIkData();
    this->ik_data_result.resolution = this->ik_data.resolution;
    this->ik_data_result.radius = this->ik_data.radius;
    this->ik_data_result.sphere_sample = this->ik_data.sphere_sample;

}

DataGenerator::~DataGenerator() {
}

void DataGenerator::data_comparator() {
    // for each sphere in the data
    for (int i = 0; i < this->ik_data.spheres.size(); i++) {
        Sphere s = this->ik_data.spheres[i];
        // create a sphere
        Sphere sphere_result;
        sphere_result.x = s.x;
        sphere_result.y = s.y;
        sphere_result.z = s.z;
        for (int k = 0; k < s.poses.size(); k++) {
            PoseOnSphere pose_on_sphere_result;
            PoseOnSphere pose_on_sphere = s.poses[k];
            pose_on_sphere_result.x = pose_on_sphere.x;
            pose_on_sphere_result.y = pose_on_sphere.y;
            pose_on_sphere_result.z = pose_on_sphere.z;
            pose_on_sphere_result.theta_x = pose_on_sphere.theta_x;
            pose_on_sphere_result.theta_y = pose_on_sphere.theta_y;
            pose_on_sphere_result.theta_z = pose_on_sphere.theta_z;
            pose_on_sphere_result.theta_w = pose_on_sphere.theta_w;

            // if p has a joint, test it
            if (pose_on_sphere.has_joints()) {
            // test the joint pose with plan
                for (int l = 0; l < pose_on_sphere.joints.size(); l++) {
                    joint j = pose_on_sphere.joints[l];
                    if (!this->robot.checkCollision(j.toVector())) {
                        // add this joint to the poseOnSphere of ik_data_result
                        pose_on_sphere_result.joints.push_back(j);
//                        ROS_INFO("Joint move done");
                        break;
                    }
                    else {
                        ROS_INFO("Joint move failed");
                    }
                }
            }
//            else {
//                // test the cartesian pose with plan
//                ROS_INFO("PointOnSphere has no joints");
//            }
            // add the poseOnSphere to the sphere
            sphere_result.add(pose_on_sphere_result);
        }
        // add the sphere to the ik_data_result
        this->ik_data_result.add(sphere_result);
    }
}



int main(int argc, char **argv) {
    ros::init(argc, argv, "data_generate");
    ros::NodeHandle n;
//    sleep 2 sec
    ros::Duration(2).sleep();
    ros::Time begin = ros::Time::now();
    ros::AsyncSpinner spinner(1);
    spinner.start();

    DataGenerator data_generator("/home/will/master_ik_data.json");
    data_generator.data_comparator();
    data_generator.ik_data_result.write_data("/home/will/master_ik_data_result.json");
    // get time
    ros::Time end = ros::Time::now();
    ros::Duration duration = end - begin;
    ROS_INFO("Total time taken: %f", duration.toSec());
    ROS_INFO("Data generation done");
    return 0;
}

File: data_generation/data_generator.py
import time
import json
from some_robotics_library import Robot, Sphere, PoseOnSphere, Joint

class DataGenerator:
    def __init__(self, filename):
        self.filename = filename
        self.robot = Robot("manipulator")
        self.ik_data = self.load_data(self.filename)
        self.ik_data_result = {
            'resolution': self.ik_data['resolution'],
            'radius': self.ik_data['radius'],
            'sphere_sample': self.ik_data['sphere_sample'],
            'spheres': []
        }

    def load_data(self, filename):
        with open(filename, 'r') as file:
            return json.load(file)

    def data_comparator(self):
        for s in self.ik_data['spheres']:
            sphere_result = Sphere(s['x'], s['y'], s['z'])
            for pose_on_sphere in s['poses']:
                pose_on_sphere_result = PoseOnSphere(
                    pose_on_sphere['x'], pose_on_sphere['y'], pose_on_sphere['z'],
                    pose_on_sphere['theta_x'], pose_on_sphere['theta_y'], pose_on_sphere['theta_z'],
                    pose_on_sphere['theta_w']
                )

                if 'joints' in pose_on_sphere:
                    for joint_data in pose_on_sphere['joints']:
                        joint = Joint.from_vector(joint_data)
                        if not self.robot.check_collision(joint.to_vector()):
                            pose_on_sphere_result.joints.append(joint)
                            break
                        else:
                            print("Joint move failed")
                sphere_result.add_pose(pose_on_sphere_result)
            self.ik_data_result['spheres'].append(sphere_result)

    def write_data(self):
        with open('/home/will/master_ik_data_result.json', 'w') as file:
            json.dump(self.ik_data_result, file, indent=2)

if __name__ == "__main__":
    data_generator = DataGenerator("/home/will/master_ik_data.json")
    time.sleep(2)
    start_time = time.time()
    data_generator.data_comparator()
    data_generator.write_data()
    end_time = time.time()
    duration = end_time - start_time
    print(f"Total time taken: {duration:.2f} seconds")
    print("Data generation done")
